lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1853,6 +1853,112 @@ LogicalResult MemoryOp::verify() {
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// SeqMemoryOp
+//===----------------------------------------------------------------------===//
+
+/// Provide meaningful names to the result values of a SeqMemoryOp.
+void SeqMemoryOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  getCellAsmResultNames(setNameFn, *this, this->portNames());
+}
+
+SmallVector<StringRef> SeqMemoryOp::portNames() {
+  SmallVector<StringRef> portNames;
+  for (size_t i = 0, e = getAddrSizes().size(); i != e; ++i) {
+    auto nameAttr =
+        StringAttr::get(this->getContext(), "addr" + std::to_string(i));
+    portNames.push_back(nameAttr.getValue());
+  }
+  portNames.append({"write_data", "write_en", "write_done", "clk", "read_data",
+                    "read_en", "read_done"});
+  return portNames;
+}
+
+SmallVector<Direction> SeqMemoryOp::portDirections() {
+  SmallVector<Direction> portDirections;
+  for (size_t i = 0, e = getAddrSizes().size(); i != e; ++i)
+    portDirections.push_back(Input);
+  portDirections.append({Input, Input, Output, Input, Output, Input, Output});
+  return portDirections;
+}
+
+SmallVector<DictionaryAttr> SeqMemoryOp::portAttributes() {
+  SmallVector<DictionaryAttr> portAttributes;
+  MLIRContext *context = getContext();
+  for (size_t i = 0, e = getAddrSizes().size(); i != e; ++i)
+    portAttributes.push_back(DictionaryAttr::get(context)); // Addresses
+
+  OpBuilder builder(context);
+  // Use a boolean to indicate this attribute is used.
+  IntegerAttr isSet = IntegerAttr::get(builder.getIndexType(), 1);
+  IntegerAttr isTwo = IntegerAttr::get(builder.getIndexType(), 2);
+  NamedAttrList writeEn, writeDone, clk, reset, readEn, readDone;
+  writeEn.append("go", isSet);
+  writeDone.append("done", isSet);
+  clk.append("clk", isSet);
+  readEn.append("go", isTwo);
+  readDone.append("done", isTwo);
+  portAttributes.append({DictionaryAttr::get(context),     // Write Data
+                         writeEn.getDictionary(context),   // Write enable
+                         writeDone.getDictionary(context), // Write done
+                         clk.getDictionary(context),       // Clk
+                         DictionaryAttr::get(context),     // Out
+                         readEn.getDictionary(context),    // Read enable
+                         readDone.getDictionary(context)}  // Read done
+  );
+  return portAttributes;
+}
+
+bool SeqMemoryOp::isCombinational() { return false; }
```

Comment:
Nit: this is the kind of thing i'd place in `extraClassDeclaration`.

Reply:
Good point, moved.